src/parser/ast/expression.rs

```diff
@@ -0,0 +1,215 @@
+use std::fmt::Debug;
+
+use super::DebugSymRef;
+
+#[derive(Clone)]
+pub enum BinaryOperator {
+    Sum,
+    Sub,
+    Mul,
+    Pow,
+    Eq,
+    NEq,
+    And,
+    Or,
+}
+
+impl BinaryOperator {
+    pub fn is_logic(&self) -> bool {
+        use BinaryOperator::*;
+        match self {
+            Eq => true,
+            NEq => true,
+            And => true,
+            Or => true,
+            Sum => false,
+            _ => false,
+        }
+    }
+
+    pub fn is_arith(&self) -> bool {
+        !self.is_logic()
+    }
+}
+
+impl Debug for BinaryOperator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Sum => write!(f, "+"),
+            Self::Sub => write!(f, "-"),
+            Self::Mul => write!(f, "*"),
+            Self::Pow => write!(f, "^"),
+            Self::Eq => write!(f, "=="),
+            Self::NEq => write!(f, "!="),
+            Self::And => write!(f, "&&"),
+            Self::Or => write!(f, "||"),
+        }
+    }
+}
+
+impl From<String> for BinaryOperator {
+    fn from(value: String) -> Self {
+        use BinaryOperator::*;
+        match value.as_str() {
+            "+" => Sum,
+            "*" => Mul,
+            "^" => Pow,
+            "==" => Eq,
+            "!=" => NEq,
+            "&&" => Or,
+            "||" => And,
+            &_ => unimplemented!(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum UnaryOperator {
+    Not,
+    Neg,
+}
+
+impl UnaryOperator {
+    pub fn is_logic(&self) -> bool {
+        use UnaryOperator::*;
+        match self {
+            Not => true,
+            Neg => false,
+        }
+    }
+
+    pub fn is_arith(&self) -> bool {
+        !self.is_logic()
+    }
+}
+
+impl Debug for UnaryOperator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use UnaryOperator::*;
+        match self {
+            Not => write!(f, "!"),
+            Neg => write!(f, "-"),
+        }
+    }
+}
+
+impl From<String> for UnaryOperator {
+    fn from(value: String) -> Self {
+        use UnaryOperator::*;
+        match value.as_str() {
+            "!" => Not,
+            "-" => Neg,
+            &_ => unimplemented!(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum Expression<F, V> {
```

Comment:
This Expression type works as a binary tree with binary operations.  So if you do `a + b + c + d` you get `a + (b + (c + d))`.
In Chiquito the expression type uses Vec for sums and muls.  Did you think about this?

Reply:
Yes, this AST reflects the grammar that is based on binary operations. We will have to implement a flattening of the chain of binaries for optimal compilation. I was think perhaps substituting lhs, rhs by an array, the parser will only produce binary, but then we can flattening in a post-process.

Reply:
I already changed it so the expressions contain an array, and i already implemeted in the compiler (next PR) to flatten the binary tree before compiling.

Reply:
Let me clarify, the target Expression contains and array instead of bianry. The compiler flatten the birary tree and produces expressions with arrays as sub expressions